[Diagnostics] SR-7445 Improve diagnostics for assignment failures

AnthonyLatsis opened the pull request:
Attempt to resolve [SR-7445](https://bugs.swift.org/browse/SR-7445).
modified messages for assignments to function calls,
modified messages for assignments to methods.
removed comment for resolved radar.

Suggestions appreciated.

Implementation proposals (`CSDiag.cpp`):
The `diagnoseSubElementFailure` function has no substantial reason to be separated from it's caller, 
`diagnoseAssignmentFailure`. Right now this separation prevents from creating a single diagnostic 
`cannot assign to %select{member|property|value}0: %1` (it is currently split into 3) and only accepts `Diag<StringRef>` (There are other diagnostics types like the aforementioned that are reasonable to be accepted. Furthermore, the current state of these two functions results in unnecessary logic separation – passing a generic diagnostic that is yet uncertain to `diagnoseSubElementFailure` can meet a need to be mutated. I propose to merge `diagnoseSubElementFailure` into `diagnoseAssignmentFailure` and tidy things up.

AnthonyLatsis committed:
[Diagnostics]

Attempt to resolve SR-7445.
modified messages for assignments to function calls,
modified messages for assignments to methods.
removed comment for resolved radar.

AnthonyLatsis committed:
removed extra line and braces

xedin reviewed (changes requested):
Since there is a new diagnostic added, could you please add couple of test-cases for `assignment_lhs_is_apply_expression`?

xedin commented:
Also please change commit message to say something like `[Diagnostics] Improve diagnostics for assignment related to calls`

AnthonyLatsis commented:
Sure. I think adding them to `immutability.swift` together with the majority of related assertions will be fine.

AnthonyLatsis committed:
added tests for assignment_lhs_is_apply_expression
eliminated redundant literal check which is always invoked before call
reverted 'cannot assign to value' for literal assignments in subexpressions

AnthonyLatsis commented:
Everything is alright, but the way the AST is generated confuses me.  The following example is a `CallExpr` that has a `DeclRefExpr`

```
func foo() {}

foo() = 5 // error: expression is not assignable: 'foo' returns immutable value
```
But if it's a method, it doesn't have one for some reason. Because of this, the diagnostic is forced to be more general:

```
class Foo { func foo() {} }

let a = Foo()

a.foo() = 7 // expression is not assignable: function call returns immutable value
```

xedin commented:
@AnthonyLatsis The reason why it's this way is because the constraint system failed to generate constraints for `a.foo() = 7` which means that it wouldn't have a change to resolve the overload yet, I think it's okay through.

xedin commented on lib/Sema/CSDiag.cpp:
```diff
@@ -3127,12 +3133,6 @@ void ConstraintSystem::diagnoseAssignmentFailure(Expr *dest, Type destTy,
   auto &TC = getTypeChecker();
 
-  // Diagnose obvious assignments to literals.
-  if (isa<LiteralExpr>(dest->getValueProvidingExpr())) {
```
I think it's still useful to diagnose stuff like `7 = 42` which this error message, but unfortunately I haven't found a test for that in `test/Sema/immutability.swift`...

AnthonyLatsis replied:
< I think it's still useful to diagnose stuff like 7 = 42
Of course, I deleted the check here because I noticed it is checked *anyways* and always in the function that calls this one. Would you like me to add a test for that?

AnthonyLatsis committed:
Complemented assigning to literal tests & reverted to 'cannot asign to value' for methods (was 'cannot assign to member')

AnthonyLatsis commented:
I found the tests in `stmt/statements.swift` and added a few more for different literals.

xedin commented on test/Sema/immutability.swift:
```diff
@@ -430,7 +430,25 @@ func QoI() {
+func func1() -> Int { return 7 }
 
+func func2() {
+    func func3() {}
```
One last nit - could you please format this to be 2 spaces like the rest of the file?

AnthonyLatsis replied:
My bad...One sec

xedin reviewed (approved):
Looks great! Once once the last spacing issue is fixed I'll get the CI running.

AnthonyLatsis committed:
removed extra tabs

AnthonyLatsis committed:
eliminated one more accidental spacing

AnthonyLatsis committed:
Update CSDiag.cpp

AnthonyLatsis commented:
It's good actually that it failed. I will fix the tests and make one more small improvement

xedin commented:
@AnthonyLatsis sure, there is `print` and `dump` methods on it, which you can call with `llvm::errs()` or `llvm::outs()`.

AnthonyLatsis commented:
Just found them. Thanks

AnthonyLatsis committed:
added highlighting, fixed & rechecked tests

added highlighting for complex expressions involving assigning to literals

AnthonyLatsis committed:
merge with minor changes commited directly on github

xedin commented:
Thank you, @AnthonyLatsis !
